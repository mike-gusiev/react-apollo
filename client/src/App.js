import React, {Component} from 'react';
import PropTypes from 'prop-types';
import ApolloClient from 'apollo-boost';
import {BrowserRouter as Router, Route, Link, Switch, Redirect} from 'react-router-dom'

import * as productDuck from './ducks/products'
import Admin from './Components/Admin'
import Customer from './Components/Customer'

import {connect} from 'react-redux';

export class App extends Component {

    componentDidMount() {
        this.props.onLoad(this.props.client)
    }

    render() {
        return (
            <Router>
                <div className="container app">
                    {this.props.products.loading ? (
                        <div>Loading...</div>
                    ) : this.props.products.error ? (
                        <div>Error: {JSON.stringify(this.props.products.error)}</div>
                    ) : (
                        <React.Fragment>
                            <nav className="nav">
                                <Link to="/admin" className='nav-link'>Admin</Link>
                                <Link to="/customer" className='nav-link'>Customer</Link>
                            </nav>
                            <Switch>
                                <Route exact path='/' render={() => (<Redirect to={'/customer'}/>)}/>
                                <Route exact path='/admin' render={() => <Admin onAdd={this.props.onAdd} client={this.props.client} />} />
                                <Route exact path='/customer' render={() => <Customer products={this.props.products}/>}/>
                            </Switch>
                        </React.Fragment>
                    )}
                </div>
            </Router>
        );
    }
}

App.propTypes = {
    client: PropTypes.instanceOf(ApolloClient),
    onAdd: PropTypes.func,
    onLoad: PropTypes.func,
    products: PropTypes.shape({
        data: PropTypes.arrayOf(PropTypes.shape({
            brand: PropTypes.shape({
                name: PropTypes.string,
                __typename: PropTypes.string
            }),
            name: PropTypes.string,
            id: PropTypes.string,
            category: PropTypes.string,
            __typename: PropTypes.string
        })),
        loading: PropTypes.bool,
        error: PropTypes.string
    })
}

export default connect(
    state => ({
        products: state.products
    }),
    dispatch => ({
        onLoad: (client) => {
            dispatch(productDuck.loadProducts(client))
        },
        onAdd: (client, vars) => {
            dispatch(productDuck.addProducts(client, vars))
        }
    })
)(App);
