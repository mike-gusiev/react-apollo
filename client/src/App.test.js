import React from 'react';
import renderer from 'react-test-renderer';
import ApolloClient from 'apollo-boost';
import {ApolloProvider} from 'react-apollo';

import {App} from './App';

test('should render component correctly', () => {
    const client = new ApolloClient({
        uri: 'http://localhost:4000/graphql'
    });
    const products = {
        products: {
            data: [
                {
                    brand: {
                        name: 'test-brand-name',
                        __typename: 'test-brand-type'
                    },
                    name: 'test-name',
                    id: 'test-id',
                    category: 'test-category',
                    __typename: 'test-typename'
                }
            ],
            loading: true,
            error: ''
        }
    };
    const tree = renderer.create(
        <ApolloProvider client={client}><App onLoad={() => {}} products={products}/></ApolloProvider>
    ).toJSON()
    expect(tree).toMatchSnapshot();
});
