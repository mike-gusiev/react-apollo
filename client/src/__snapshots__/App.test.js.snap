// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component correctly 1`] = `
<div
  className="container app"
>
  <nav
    className="nav"
  >
    <a
      className="nav-link"
      href="/admin"
      onClick={[Function]}
    >
      Admin
    </a>
    <a
      className="nav-link"
      href="/customer"
      onClick={[Function]}
    >
      Customer
    </a>
  </nav>
  <h2>
    Costumer
  </h2>
  <hr />
  <ul
    className="product-list"
  />
</div>
`;
